### PR TITLE
add custom nodes implementation

### PR DESCRIPTION
**Pull request by angad-k**

### CODE REVIEW

File: ProcSDF/GUI/Inspector.cpp [hunk 1]
```diff
@@ -94,6 +96,24 @@ void Inspector::drawRenderingSettings()
...
+	{
+		std::pair<bool, std::string> l_pickFileResult = OS::pickFile();
+		if (l_pickFileResult.first)
+		{
+			std::string l_filePath = l_pickFileResult.second;
+			CustomNode* l_customNode = new CustomNode(l_filePath);
+			if (l_customNode->isMalformed())
+			{
+				delete(l_customNode);
+			}
```
Comment (ayushgupta138) [inline ProcSDF/GUI/Inspector.cpp:109]: displaying some errors would be helpful here

File: ProcSDF/GUI/Inspector.cpp [hunk 2]
```diff
@@ -94,6 +96,24 @@ void Inspector::drawRenderingSettings()
...
 void Inspector::drawNodeGraphSettings()
 {
+	if (ImGui::Button("Add custom node"))
+	{
+		std::pair<bool, std::string> l_pickFileResult = OS::pickFile();
+		if (l_pickFileResult.first)
+		{
+			std::string l_filePath = l_pickFileResult.second;
+			CustomNode* l_customNode = new CustomNode(l_filePath);
+			if (l_customNode->isMalformed())
```
Comment (ayushgupta138) [inline ProcSDF/GUI/Inspector.cpp:106]: moreover, we are not checking if the file submitted has a compilation error. which leads to stalling of the program if there is one. We need to fix this in the next pr.

File: ProcSDF/GUI/Nodes/CustomNode.cpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+CustomNode::CustomNode(std::string p_filePath)
+{
+	std::string l_fileContent = OS::fetchFileContent(p_filePath);
+	std::vector<std::string> l_lines = tokenizeWithDelimiters(l_fileContent, "\n");
+	for (std::string line : l_lines)
+	{
+		std::vector<std::string> l_tokens = tokenizeWithDelimiters(line, "\\s+");
+
+		if (l_tokens[0] == "//")
```
Comment (ayushgupta138) [inline ProcSDF/GUI/Nodes/CustomNode.cpp:28]: this way of passing metadata is not parsing-friendly and results in ugly code, we need to change this in the future. It's okay in this pr, though.

File: ProcSDF/GUI/Nodes/CustomNode.cpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+					break;
+				}
+				m_nodeName = l_tokens[2];
+			}
+			else if (l_tokens[1] == "input_pins")
+			{
+				for (int i = 2; i < l_tokens.size(); i++)
+				{
+					m_inputPins.push_back(l_tokens[i]);
+				}
```
Comment (ayushgupta138) [inline ProcSDF/GUI/Nodes/CustomNode.cpp:66]: use std::insert here, m_inputPins.insert(m_inputPins.end(), l_tokens.begin() + 2, l_tokens.end());

File: ProcSDF/GUI/Nodes/CustomNode.cpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+			}
+			else if (l_tokens[1] == "vec3_params")
+			{
+				for (int i = 2; i < l_tokens.size(); i++)
+				{
+					m_inputFloat3Labels.push_back(l_tokens[i]);
+				}
```
Comment (ayushgupta138) [inline ProcSDF/GUI/Nodes/CustomNode.cpp:73]: use insert

File: ProcSDF/GUI/Nodes/CustomNode.cpp [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+			}
+			else if (l_tokens[1] == "float_params")
+			{
+				for (int i = 2; i < l_tokens.size(); i++)
+				{
+					m_inputFloatLabels.push_back(l_tokens[i]);
+				}
```
Comment (ayushgupta138) [inline ProcSDF/GUI/Nodes/CustomNode.cpp:80]: use insert

File: ProcSDF/Rendering/ShaderGenerator.cpp [hunk 7]
```diff
@@ -37,6 +38,15 @@ void ShaderGenerator::computeUniforms()
 	}
 }
 
+void ShaderGenerator::appendCustomFunctions(std::string &p_shaderString)
+{
+	std::vector<std::string> l_customNodeFileContents = NodeGraph::getSingleton()->getCustomNodeFileContents();
+	for (std::string l_fileContent : l_customNodeFileContents)
+	{
+		p_shaderString.append(l_fileContent);
+	}
```
Comment (ayushgupta138) [inline ProcSDF/Rendering/ShaderGenerator.cpp:47]: use insert

File: ProcSDF/GUI/Inspector.cpp (hunk 1 above)
Comment (angad-k) [reply ProcSDF/GUI/Inspector.cpp:109]: we are outputting an error wherever we set the isMalformed thingie.

File: ProcSDF/GUI/Nodes/CustomNode.cpp (hunk 3 above)
Comment (angad-k) [reply ProcSDF/GUI/Nodes/CustomNode.cpp:28]: explain? we could just get it done with in this PR..

File: ProcSDF/GUI/Inspector.cpp (hunk 2 above)
Comment (angad-k) [reply ProcSDF/GUI/Inspector.cpp:106]: #34 opened

File: ProcSDF/GUI/Nodes/CustomNode.cpp (hunk 3 above)
Comment (ayushgupta138) [reply ProcSDF/GUI/Nodes/CustomNode.cpp:28]: For example, we could use JSON for metadata passing, libraries like jsoncpp can reduce the logic to < 10 lines of code. That would require additional work of integrating a library. Your call if you want to do it in this pr.

Comment (angad-k) [reply ProcSDF/GUI/Nodes/CustomNode.cpp:28]: Nevermind, merge

**Review by ayushgupta138 — APPROVED** [review]

**Pull request merged**